lib/Monads/nondet/Nondet_While_Loop_Rules.thy

```diff
@@ -498,32 +508,43 @@ lemma empty_fail_whileM[empty_fail_cond, intro!, wp]:
 
 lemma whileLoop_results_bisim:
   assumes base: "(a, b) \<in> whileLoop_results C B"
-  and vars1: "Q = (case a of Some (r, s) \<Rightarrow> Some (rt r, st s) | _ \<Rightarrow> None)"
-  and vars2: "R = (case b of Some (r, s) \<Rightarrow> Some (rt r, st s) | _ \<Rightarrow> None)"
-  and inv_init: "case a of Some (r, s) \<Rightarrow> I r s | _ \<Rightarrow> True"
-  and inv_step: "\<And>r s r' s'. \<lbrakk> I r s; C r s; (r', s') \<in> fst (B r s) \<rbrakk> \<Longrightarrow> I r' s'"
-  and cond_match: "\<And>r s. I r s \<Longrightarrow> C r s = C' (rt r) (st s)"
-  and fail_step: "\<And>r s. \<lbrakk>C r s; snd (B r s); I r s\<rbrakk>
+    and vars1: "Q = (case a of Some (r, s) \<Rightarrow> Some (rt r, st s) | _ \<Rightarrow> None)"
+    and vars2: "R = (case b of Some (r, s) \<Rightarrow> Some (rt r, st s) | _ \<Rightarrow> None)"
+    and inv_init: "case a of Some (r, s) \<Rightarrow> I r s | _ \<Rightarrow> True"
+    and inv_step: "\<And>r s r' s'. \<lbrakk> I r s; C r s; (r', s') \<in> fst (B r s) \<rbrakk> \<Longrightarrow> I r' s'"
+    and cond_match: "\<And>r s. I r s \<Longrightarrow> C r s = C' (rt r) (st s)"
+    and fail_step: "\<And>r s. \<lbrakk>C r s; snd (B r s); I r s\<rbrakk>
                          \<Longrightarrow> (Some (rt r, st s), None) \<in> whileLoop_results C' B'"
-  and refine: "\<And>r s r' s'. \<lbrakk> I r s; C r s; (r', s') \<in>  fst (B r s) \<rbrakk>
+    and refine: "\<And>r s r' s'. \<lbrakk> I r s; C r s; (r', s') \<in>  fst (B r s) \<rbrakk>
                             \<Longrightarrow> (rt r', st s') \<in> fst (B' (rt r) (st s))"
   shows "(Q, R) \<in> whileLoop_results C' B'"
   apply (subst vars1)
   apply (subst vars2)
   apply (insert base inv_init)
-  apply (induct rule: whileLoop_results.induct)
+proof (induct rule: whileLoop_results.induct)
```

Comment:
Same as above, the AFP (and proof style in general) will frown on `apply` followed by `proof` (we have a few instances of that in other proofs, they are not good examples to follow :-)).

This one is a bit more complex. My first instinct would be to try something with `defines`, but I can't quite see in the diff if that makes sense and which names `Q` and `R` mention on the right-hand side.

Reply:
I'm pretty much done for the week so will have to continue this PR review next week, but: I very much object to `proof` in the middle of `apply` scripts. We have some historical leftovers (mainly from SimonW), but we should not introduce more of that.

Reply:
I've gone with making them `defines [simp]:`. Seeing how the previous proof unfolded them straight away, I'm assuming that they were introduced mainly to improve readability.
I did also have to parameterise them to get the induction to apply properly.

Reply:
Unfortunately, that didn't actually work, I've realised that the intention for how it was previously written was to make the rule easy to apply where it's used.
I've now split it into two lemmas; one helper that uses `defines` and does the induction, and a second that has the same lemma statement as before my changes.